interpreter.go

```diff
@@ -20,159 +20,70 @@ import (
-	// TODO(dcunnin): Include stacktrace.
-	return err.Msg
-}
-
-// Values and state
-
-type bindingFrame map[identifier]*thunk
-
-type value interface {
-}
-
-type valueString struct {
-	value string
-}
-
-func makeValueString(v string) *valueString {
-	return &valueString{value: v}
-}
-
-type valueBoolean struct {
-	value bool
-}
-
-func makeValueBoolean(v bool) *valueBoolean {
-	return &valueBoolean{value: v}
-}
-
-type valueNumber struct {
-	value float64
-}
-
-func makeValueNumber(v float64) *valueNumber {
-	return &valueNumber{value: v}
-}
-
-// TODO(dcunnin): Maybe intern values null, true, and false?
-type valueNull struct {
-}
-
-func makeValueNull() *valueNull {
-	return &valueNull{}
-}
-
-type thunk struct {
-	content  value // nil if not filled
-	name     identifier
+// TODO(sbarzowski) use it as a pointer in most places b/c it can sometimes be shared
+// for example it can be shared between array elements and function arguments
+type environment struct {
+	sb selfBinding
+
+	// Bindings introduced in this frame. The way previous bindings are treated
+	// depends on the type of a frame.
+	// If isCall == true then previous bindings are ignored (it's a clean
+	// environment with just the variables we have here).
+	// If isCall == false then if this frame doesn't contain a binding
+	// previous bindings will be used.
 	upValues bindingFrame
-	self     value
-	offset   int
-	body     astNode
 }
 
-func makeThunk(name identifier, self value, offset int, body astNode) *thunk {
-	return &thunk{
-		name:   name,
-		self:   self,
-		offset: offset,
-		body:   body,
+func makeEnvironment(upValues bindingFrame, sb selfBinding) environment {
+	return environment{
+		upValues: upValues,
+		sb:       sb,
 	}
 }
 
-func (t *thunk) fill(v value) {
-	t.content = v
-	t.self = nil
-	t.upValues = make(bindingFrame) // clear the map
-}
-
-func (t *thunk) filled() bool {
-	return t.content != nil
+func callFrameToTraceFrame(frame *callFrame) TraceFrame {
+	return traceElementToTraceFrame(frame.trace)
 }
 
-type valueArray struct {
-	elements []*thunk
-}
-
-func makeValueArray(elements []*thunk) *valueArray {
-	return &valueArray{
-		elements: elements,
+func (i *interpreter) getCurrentStackTrace(additional *TraceElement) []TraceFrame {
+	var result []TraceFrame
+	for _, f := range i.stack.stack {
+		if f.isCall {
+			result = append(result, callFrameToTraceFrame(f))
+		}
 	}
+	if additional != nil {
+		result = append(result, traceElementToTraceFrame(additional))
+	}
+	return result
 }
 
-type valueClosure struct {
-	upValues bindingFrame
-}
-
-type valueSimpleObjectField struct {
-	hide astObjectFieldHide
-	body astNode
-}
-
-type valueSimpleObjectFieldMap map[string]valueSimpleObjectField
-
-type valueSimpleObject struct {
-	upValues bindingFrame
-	fields   valueSimpleObjectFieldMap
-	asserts  []astNode
-}
+type callFrame struct {
+	// True if it switches to a clean environment (function call or thunk)
+	// False otherwise, e.g. for local
```

Comment:
the association between locals and thunks makes this a bit unclear, use another ast for hte example, like the conditional